File: protocol/protocol.py
"""
protocol/protocol.py — wire format constants, shared with protocol.h + protocol.rs.

Frame: [magic:4][payload_size:4][type_tag:4][payload_body...]
"""
import struct
from enum import IntEnum

MAGIC: int = 0x4D415246
FRAME_HEADER_SIZE: int = 12  # magic(4) + size(4) + type_tag(4)

DEFAULT_TCP_PORT: int = 9999
DEFAULT_PIPE_NAME: str = "tictactoe_stream"

HEADER_STRUCT = struct.Struct("<III")  # magic, size, type_tag


class PayloadType(IntEnum):
    NONE = 0
    BGRA_FRAME = 1
    H264_STREAM = 2
    CONTROL_MSG = 3
    CAPABILITIES = 4


def build_header(payload_size: int, type_tag: PayloadType) -> bytes:
    return HEADER_STRUCT.pack(MAGIC, payload_size, int(type_tag))


def parse_header(data: bytes) -> tuple[int, PayloadType] | None:
    magic, size, tag = HEADER_STRUCT.unpack(data)
    if magic != MAGIC:
        return None
    try:
        return size, PayloadType(tag)
    except ValueError:
        return None

File: protocol/protocol.rs
// protocol/protocol.rs — wire format constants, shared with protocol.h + protocol.py.
//
// Frame: [magic:4][payload_size:4][type_tag:4][payload_body...]
//
// This module has zero dependencies on transport or payload modules.
// Some items are unused within the Rust binary but kept for cross-language protocol parity.

#![allow(dead_code)]

pub const MAGIC: u32 = 0x4D415246;
pub const DEFAULT_TCP_PORT: u16 = 9999;

/// Payload type tags — match protocol.h PayloadType enum.
#[repr(u32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PayloadType {
    BgraFrame   = 1,
}

/// Build a frame header: [magic:4][size:4][type:4]
pub fn build_header(payload_size: u32, type_tag: PayloadType) -> [u8; 12] {
    let mut h = [0u8; 12];
    h[0..4].copy_from_slice(&MAGIC.to_le_bytes());
    h[4..8].copy_from_slice(&payload_size.to_le_bytes());
    h[8..12].copy_from_slice(&(type_tag as u32).to_le_bytes());
    h
}
